### DISCUSSION TITLE
How to evaluate a variable at a particular position/time

### DISCUSSION THREAD

**Discussion by brosaplanella** [pybamm-team/PyBaMM#1295] [Q&A] [answered]
Here is a quite common question:

> How can I extract the model variable at a particular position of the electrode, `x`? For example:
```python3
pybamm.dynamic_plot(
    simulation,
    output_variables =[
        "Negative particle surface concentration [mol.m-3]",
    ],
)
```

> This plot is giving evolution of the surface concentration in the negative electrode with time. Can I pick, for example, the value `x = 1e-5` and plot the concentration at that point vs time?

**Comment by brosaplanella** [comment, marked as answer]
You can do that using post-processed variables after you have solved your `simulation`. Then, you can define your post-processed variable as
```python3
c_n_surf = simulation.solution["Negative particle surface concentration [mol.m-3]"]
```
and you can find all the variables you can call in [this notebook](https://colab.research.google.com/github/pybamm-team/PyBaMM/blob/master/examples/notebooks/Getting%20Started/Tutorial%203%20-%20Basic%20plotting.ipynb).

This returns an object `c_n_surf` which is an interpolant. Then, we can call that object and evaluate it at any point by doing
```python3
c_n_surf(x = 1e-5, t = 100)
```

Because this is a surface concentration, it only depends on `x` and `t`. If we chose the concentration in the particles it would depend on `r`, `x` and `t` and thus we would have to evaluate it for the three variables. Note that you can pass time as a vector to evaluate the variable at multiple times.

**Reply by ssingh-ipa** [reply]
Could you please elaborate the x and t here. Is this the time in [s], [min] or [h]? I am guessing x is the μm molar concentration or the distance between electrodes. What is correct?  

When I save this data in a .mat file, it looks like the following image
![Negative particle surface concentration](https://user-images.githubusercontent.com/65173961/120518490-ee635300-c3d1-11eb-8d36-ba8761c63f63.PNG)
Current, voltage and time are 1D, could you please explain the 2D aspect of the remaining output variables such as 'Negative particle surface concentration [mol.m-3]' or 'Electrolyte concentration [mol.m-3]'.

**Reply by valentinsulzer** [reply]
> Could you please elaborate the x and t here. Is this the time in [s], [min] or [h]? I am guessing x is the μm molar concentration or the distance between electrodes. What is correct?

`t` is time in seconds and `x` is distance in metres from the anode current collector. Same as shows up on the x axis when you plot using `dynamic_plot`. 

> Current, voltage and time are 1D, could you please explain the 2D aspect of the remaining output variables such as 'Negative particle surface concentration [mol.m-3]' or 'Electrolyte concentration [mol.m-3]'.

Some variables such as 'Negative particle surface concentration [mol.m-3]' or 'Electrolyte concentration [mol.m-3]' depend on both space and time. In your screenshot each column is a point in time and each row is a point in space. If you also save "Time [s]" and "x [m]" you will get the corresponding time and space at which those points are evaluated.

**Reply by ssingh-ipa** [reply]
Thank you. There are a total of 539 model variables and only 98 parameter_values for any chosen parameter set (from the various citations). Are all the model variables parameterized by default when a model (such as DFN model) is executed, or are the parameter_values enough to execute a model? I am only wondering the use of the remaining 441 variables. How can I realize the relevant set of model results without having to explore through all the 539 model variables.

**Reply by valentinsulzer** [reply]
The most relevant variables are the ones plotted by default in the standard examples. Beyond that it is up to you which variables you want to look at, which requires some understanding of the problem you are trying to solve. I recommend you go through our examples (e.g. getting started) to see the difference between a parameter and a variable
